libslide/src/evaluator_rules.rs

```diff
@@ -109,36 +112,64 @@ impl Transformer<Expr, Expr> for BuiltRule {
+            cache: &mut HashMap<u64 /* ptr address */, Rc<Expr>>,
+        ) -> Rc<Expr> {
+            if let Some(result) = cache.get(&hash(target.as_ref())) {
+                return Rc::clone(result);
             }
-            .into(),
-            Expr::Parend(expr) => Expr::Parend(self.transform(*expr).into()),
-            Expr::Braced(expr) => Expr::Braced(self.transform(*expr).into()),
-        };
-
-        let replacements = match match_rule(self.from.clone(), partially_transformed.clone()) {
-            Some(repl) => repl,
-            // Could not match the rule on the top-level of the expression; return the partially
-            // transformed expression.
-            None => return partially_transformed,
-        };
-
-        // If the rule was matched on the expression, we have replacements for rule patterns ->
-        // target subexpressions. Apply the rule by transforming the rule's RHS with the replacements.
-        replacements.transform(self.to.clone())
+
+            // First, apply the rule recursively on the target's subexpressions.
+            let partially_transformed = match target.as_ref() {
+                Expr::Const(_) => Rc::clone(&target),
+                Expr::Var(_) => Rc::clone(&target),
+                Expr::BinaryExpr(binary_expr) => Expr::BinaryExpr(BinaryExpr {
+                    op: binary_expr.op,
+                    lhs: transform(rule, Rc::clone(&binary_expr.lhs), cache),
+                    rhs: transform(rule, Rc::clone(&binary_expr.rhs), cache),
+                })
+                .into(),
+                Expr::UnaryExpr(unary_expr) => Expr::UnaryExpr(UnaryExpr {
+                    op: unary_expr.op,
+                    rhs: transform(rule, Rc::clone(&unary_expr.rhs), cache),
+                })
+                .into(),
+                Expr::Parend(expr) => {
+                    let inner = transform(rule, Rc::clone(expr), cache);
+                    Expr::Parend(inner).into()
+                }
+                Expr::Braced(expr) => {
+                    let inner = transform(rule, Rc::clone(expr), cache);
+                    Expr::Braced(inner).into()
+                }
+            };
+
+            let replacements = match match_rule(
+                rule.from.clone().into(),
+                (*partially_transformed).clone().into(),
+            ) {
+                Some(repl) => repl,
+                // Could not match the rule on the top-level of the expression; return the partially
+                // transformed expression.
+                None => return fill(cache, target, partially_transformed),
+            };
+
+            // If the rule was matched on the expression, we have replacements for rule patterns ->
+            // target subexpressions. Apply the rule by transforming the rule's RHS with the
+            // replacements.
+            let transformed = replacements.transform(rule.to.clone().into());
+            fill(cache, target, transformed)
```

Comment:
I think this match statement can be conslidated with the match expression above. so that we just return a tranformed repl.

Reply:
How? First we need to apply the rule recursively (first match expression), then rule pattern matching.

Reply:
```
return match match_rule (rule.from.clone().into(), (*partially_transformed).clone().into() {
Some(repl) => let tranformed = repl.tranform(rule.to.clone().into()); fill(cache, target, transformed)
None => fill(cache, target, partially_transformed)
}
```

Reply:
something like that. what do you think

Reply:
i see now. done, thanks